onnx_tf/handlers/backend/loop.py

```diff
@@ -0,0 +1,156 @@
+    # the initial value of scan_outputs for the second loop
+    def run_first_loop(cond, v):
+      input_values = {}
+      input_values[body.input[0].name] = M
+      input_values[body.input[1].name] = cond
+      for i in range(2, len(body.input)):
+        input_values[body.input[i].name] = v[i - 2]
+      tensor_dict = onnx_tf.backend.onnx_graph_to_tensorflow_ops(
+          graph_def=body, input_values=input_values, opset=current_opset)
+      outputs = [tensor_dict[output.name] for output in body.output]
+      scan_outputs = [
+          tf.expand_dims(outputs[i], 0)
+          for i in range(scan_outputs_start_index, len(outputs))
+      ]
+      return outputs[0], outputs[1:scan_outputs_start_index], scan_outputs
+
+    def run_no_op(cond, v):
+      # The true_fn and false_fn in tf.cond must have the same
+      # non-zero number and type of outputs. Therefore need to
+      # set the dummy scan_outputs with the correct dtype.
+      scan_outputs = [
+          init_value_dict[body.output[i].type.tensor_type.elem_type]
+          for i in range(scan_outputs_start_index, len(body.output))
+      ]
+      return cond, v, scan_outputs
+
+    def run_subgraph(cond, v, scan_outputs):
+      input_values = {}
+      input_values[body.input[0].name] = M
+      input_values[body.input[1].name] = cond
+      for i in range(2, len(body.input)):
+        input_values[body.input[i].name] = v[i - 2]
+      tensor_dict = onnx_tf.backend.onnx_graph_to_tensorflow_ops(
+          graph_def=body, input_values=input_values, opset=current_opset)
+      outputs = [tensor_dict[output.name] for output in body.output]
+      for i in range(scan_outputs_start_index, len(outputs)):
+        val_i = tf.expand_dims(outputs[i], 0)
+        sout_index = i - scan_outputs_start_index
+        scan_outputs[sout_index] = tf.concat([scan_outputs[sout_index], val_i],
+                                             0)
+      return outputs[0], outputs[1:scan_outputs_start_index], scan_outputs
+
+    # for loop
+    if M is not None and cond is None:
+      _, v, scan_outputs = tf.cond(tf.greater(M, 0),
+                                   lambda: run_first_loop("", v_initial),
+                                   lambda: run_no_op("", v_initial))
+      scan_outputs_shapes = [tf.TensorShape(None) for o in scan_outputs]
+      M = tf.cast(M, tf.int32)
+      condition = lambda cond, v, scan_outputs: True
+      _, v_final, scan_outputs = tf.while_loop(cond=condition,
+                                               body=run_subgraph,
+                                               loop_vars=["", v, scan_outputs],
+                                               shape_invariants=[
+                                                   tf.TensorShape(None),
+                                                   v_shapes, scan_outputs_shapes
+                                               ],
+                                               maximum_iterations=M - 1)
```

Comment:
I wonder if M can be 0 coming in. If that never could be the case, do we still need the tf.cond with run_no_op? If M can be 0, would the rest of code work as expected?

Reply:
@chinhuang007 M = 0 or cond = False are both valid inputs, that will make the loop doesn't even need to run once. 
I modify my code and verify that it can handle those corner cases, may you review it again please.